model/base_code_model.py

```diff
@@ -0,0 +1,158 @@
+
+    # ===== DATALOADERS BLOCK =====
+
+    def train_dataloader(self) -> DataLoader:
+        dataloader, n_samples = create_dataloader(
+            self.config.train_data_path,
+            self.config.max_context,
+            self.config.random_context,
+            self.config.shuffle_data,
+            self.config.batch_size,
+            self.num_workers,
+        )
+        print(f"approximate number of steps for train is {ceil(n_samples / self.config.batch_size)}")
+        return dataloader
+
+    def val_dataloader(self) -> DataLoader:
+        dataloader, n_samples = create_dataloader(
+            self.config.val_data_path,
+            self.config.max_context,
+            False,
+            False,
+            self.config.test_batch_size,
+            self.num_workers,
+        )
+        print(f"approximate number of steps for val is {ceil(n_samples / self.config.test_batch_size)}")
+        return dataloader
+
+    def test_dataloader(self) -> DataLoader:
+        dataloader, n_samples = create_dataloader(
+            self.config.test_data_path,
+            self.config.max_context,
+            False,
+            False,
+            self.config.test_batch_size,
+            self.num_workers,
+        )
+        print(f"approximate number of steps for test is {ceil(n_samples / self.config.test_batch_size)}")
+        return dataloader
+
+    # ===== STEP =====
+
+    def transfer_batch_to_device(self, batch: PathContextBatch, device: torch.device) -> PathContextBatch:
+        # Dict str -> torch.Tensor [seq length; batch size * n_context]
+        for k in batch.context:
+            batch.context[k] = batch.context[k].to(self.device)
+        # [seq length; batch size]
+        batch.labels = batch.labels.to(self.device)
+        return batch
+
+    def training_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
+        # [seq length; batch size; vocab size]
+        logits = self(batch.context, batch.contexts_per_label, batch.labels.shape[0], batch.labels)
+        loss = self._calculate_loss(logits, batch.labels)
+        log = {"train/loss": loss}
+        with torch.no_grad():
+            statistic = self._compute_metrics(logits, batch.labels)
+
+        log.update(statistic.calculate_metrics(group="train"))
+        progress_bar = self._get_progress_bar(log, "train")
+
+        return {"loss": loss, "log": log, "progress_bar": progress_bar, "statistic": statistic}
+
+    def validation_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
+        # [seq length; batch size; vocab size]
+        logits = self(batch.context, batch.contexts_per_label, batch.labels.shape[0])
+        loss = self._calculate_loss(logits, batch.labels)
+        with torch.no_grad():
+            statistic = self._compute_metrics(logits, batch.labels)
+
+        return {"val_loss": loss, "statistic": statistic}
+
+    def test_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
```

Comment:
I'm not sure about this. It can stay here, but I think if we make it abstract it will be more accurate.